src/base-api-client.ts

```diff
@@ -0,0 +1,99 @@
+import { ApiRequestHandler } from './api-request-handler';
+
+export type ServiceApi = 'topic_extraction' | 'sentiment_analysis';
+
+/**
+ * Base client implementation. Intended to be extended by a specific client per API
+ */
+export class BaseApiClient<TJob, TSubmitOptions, TResult, TResultOptions> {
+    apiHandler: ApiRequestHandler;
+
+    /**
+     * @param accessToken Access token used to validate API requests
+     * @param serviceApi Type of api service
+     * @param version (optional) version of the API to be used
+     */
+    constructor (accessToken: string, serviceApi: ServiceApi, version = 'v1') {
+        this.apiHandler = new ApiRequestHandler(`https://api.rev.ai/${serviceApi}/${version}/`, accessToken);
+    }
+
+    /**
+     * Get information about a specific job
+     * @param id Id of job whose details are to be retrieved
+     * @returns Job details
+     */
+    protected async getJobDetails(id: string): Promise<TJob> {
+        return await this.apiHandler.makeApiRequest<TJob>('get', `/jobs/${id}`, {}, 'json');
+    }
+
+    /**
+     * Get a list of jobs submitted within the last 30 days in reverse chronological order 
+     * (last submitted first) up to the provided limit number of jobs per call. Pagination is supported via passing
+     * the last job id from previous call into starting_after.
+     * @param limit (optional) maximum number of jobs to retrieve, default is 100
+     * @param startingAfter (optional) returns only jobs created after the job with this id, exclusive
+     * @returns List of job details
+     */
+    protected async getListOfJobs(limit?: number, startingAfter?: string): Promise<TJob[]> {
```

Comment:
Another reason I'm not a fan of TResultOptions. It makes our interfaces inconsistent with this method. In some cases the options are nullable parameters, in others they're objects